src/main/java/com/metamx/common/CompressionUtils.java

```diff
@@ -0,0 +1,348 @@
+        new FilterInputStream(in)
+        {
+          @Override
+          public int available()
+          {
+            // Hack. Docs say available() should return an estimate, so we estimate about 1KB to work around available == 0 bug in GZIPInputStream
+            return 1 << 10;
+          }
+        }
+    );
+  }
+
+  public static long gunzip(InputStream in, OutputStream out) throws IOException
+  {
+    try (GZIPInputStream gzipInputStream = gzipInputStream(in)) {
+      return ByteStreams.copy(gzipInputStream, out);
+    }
+    finally {
+      out.close();
+    }
+  }
+
+  /**
+   * A gunzip function to store locally
+   *
+   * @param in          The factory to produce input streams
+   * @param outFile     The file to store the result into
+   * @param shouldRetry A predicate to indicate if the Throwable is recoverable
+   *
+   * @return The count of bytes written to outFile
+   */
+  public static FileUtils.FileCopyResult gunzip(
+      final ByteSource in,
+      final File outFile,
+      Predicate<Throwable> shouldRetry
+  )
+  {
+    return FileUtils.retryCopy(
+        new ByteSource()
+        {
+          @Override
+          public InputStream openStream() throws IOException
+          {
+            return gzipInputStream(in.openStream());
+          }
+        },
+        shouldRetry,
+        outFile,
+        DEFAULT_RETRY_COUNT
+    );
+  }
+
+
+  public static FileUtils.FileCopyResult gunzip(final ByteSource in, File outFile)
+  {
+    return gunzip(in, outFile, FileUtils.MD5ERROR);
+  }
+
+  public static long gzip(InputStream inputStream, OutputStream out) throws IOException
+  {
+    try (GZIPOutputStream outputStream = new GZIPOutputStream(out)) {
+      return ByteStreams.copy(inputStream, outputStream);
+    }
+    finally {
+      out.close();
+      inputStream.close();
+    }
+  }
+
+  public static FileUtils.FileCopyResult gzip(final File inFile, final File outDir, Predicate<Throwable> shouldRetry)
+      throws IOException
+  {
+    final File outFile = Paths.get(outDir.getAbsolutePath(), inFile.getName() + GZ_SUFFIX).toFile();
+    FileUtils.retryCopy(
+        new ByteSource()
+        {
+          @Override
+          public InputStream openStream() throws IOException
+          {
+            return new FileInputStream(inFile);
+          }
+        },
+        shouldRetry,
+        new ByteSink()
+        {
+          @Override
+          public OutputStream openStream() throws IOException
+          {
+            return new GZIPOutputStream(new FileOutputStream(outFile));
+          }
+        },
+        DEFAULT_RETRY_COUNT
+    );
+    return FileUtils.FileCopyResult.buildResult(outFile.toPath(), "MD5");
+  }
+
+
+  public static FileUtils.FileCopyResult gzip(final File inFile, final File outDir) throws IOException
+  {
+    return gzip(inFile, outDir, FileUtils.MD5ERROR);
+  }
+
+  public static Boolean isZip(String fName)
+  {
+    if (Strings.isNullOrEmpty(fName)) {
```

Comment:
Since this is a library function I tried to
1. Make the inputs work for crazy inputs, not just the normal use case
2. Make the logic as obvious as possible to anyone viewing the source.

I could also make it a big boolean expression on one line, but I don't like having logic that takes me more than 1 second to figure out exactly what its doing.